### ISSUE TITLE
Changing a VM's state from "Stopped" to "Running" after provisioning it "Stopped" causes an error when trying to create a DNS record for the VM

### ISSUE THREAD

**Issue by simu** [exoscale/terraform-provider-exoscale#101]
Given the following Terraform configuration

```
terraform {
  required_version = ">= 0.14"
  required_providers {
    exoscale = {
      source  = "exoscale/exoscale"
      version = "~> 0.23"
    }
  }
}

variable "ssh_key" {
  type        = string
  description = "Name of an already present SSH Keypair"
}

variable "node_state" {
  type    = string
  default = "Stopped"
}

variable "domain" {
  type        = string
  description = "Domain name to create in Exoscale DNS"
}

data "exoscale_compute_template" "ubuntu2004" {
  zone = "ch-dk-2"
  name = "Linux Ubuntu 20.04 LTS 64-bit"
}

resource "exoscale_domain" dom {
  name = var.domain
}

resource "exoscale_compute" "vm" {
  display_name       = "test-vm"
  hostname           = "test-vm"
  key_pair           = var.ssh_key
  zone               = "ch-dk-2"
  template_id        = data.exoscale_compute_template.ubuntu2004.id
  size               = "Micro"
  disk_size          = 10
  state              = var.node_state
}

resource "exoscale_domain_record" "vm" {
  domain      = exoscale_domain.dom.id
  name        = "test-vm"
  ttl         = 60
  record_type = "A"
  content     = exoscale_compute.vm.ip_address
}
```

creating the DNS A record for the VM in "Stopped" state fails with:

```
$ export EXOSCALE_API_KEY=EXO....
$ export EXOSCALE_API_SECRET=....
$ terraform apply -var ssh_key=simon.gerber@vshn2021 -var domain=c-exo-ocp4-test.appuio-beta.ch

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # exoscale_compute.vm will be created
  + resource "exoscale_compute" "vm" {
      + affinity_group_ids = (known after apply)
      + affinity_groups    = (known after apply)
      + disk_size          = 10
      + display_name       = "test-vm"
      + gateway            = (known after apply)
      + hostname           = "test-vm"
      + id                 = (known after apply)
      + ip4                = true
      + ip6                = false
      + ip6_address        = (known after apply)
      + ip6_cidr           = (known after apply)
      + ip_address         = (known after apply)
      + key_pair           = "simon.gerber@vshn2021"
      + name               = (known after apply)
      + password           = (sensitive value)
      + security_group_ids = (known after apply)
      + security_groups    = (known after apply)
      + size               = "Micro"
      + state              = "Stopped"
      + tags               = (known after apply)
      + template           = (known after apply)
      + template_id        = "985a353f-2e5a-4b67-ab9f-13f12d5a69f7"
      + user_data_base64   = (known after apply)
      + username           = (known after apply)
      + zone               = "ch-dk-2"
    }

  # exoscale_domain.dom will be created
  + resource "exoscale_domain" "dom" {
      + auto_renew = (known after apply)
      + expires_on = (known after apply)
      + id         = (known after apply)
      + name       = "c-exo-ocp4-test.appuio-beta.ch"
      + state      = (known after apply)
      + token      = (known after apply)
    }

  # exoscale_domain_record.vm will be created
  + resource "exoscale_domain_record" "vm" {
      + content     = (known after apply)
      + domain      = (known after apply)
      + hostname    = (known after apply)
      + id          = (known after apply)
      + name        = "test-vm"
      + prio        = (known after apply)
      + record_type = "A"
      + ttl         = 60
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

exoscale_domain.dom: Creating...
exoscale_compute.vm: Creating...
exoscale_domain.dom: Creation complete after 2s [id=c-exo-ocp4-test.appuio-beta.ch]
exoscale_compute.vm: Creation complete after 5s [id=eabb91be-5388-4bbd-9bcc-3e16d062b428]
exoscale_domain_record.vm: Creating...

Error: dns error: Validation failed (content: can't be blank, is an invalid IPv4 address)

  on main.tf line 21, in resource "exoscale_domain_record" "vm":
  21: resource "exoscale_domain_record" "vm" {


```

This can be worked around by ensuring the domain record is only created when the requested node state is "Running":

```
resource "exoscale_domain_record" "vm" {
  count       = var.node_state == "Running" ? 1 : 0
  domain      = exoscale_domain.dom.id
  name        = "test-vm"
  ttl         = 60
  record_type = "A"
  content     = exoscale_compute.vm.ip_address
}
```

but then a subsequent `terraform apply` with an explicit request of `node_state=Running` fails with:

```
$ terraform apply -var ssh_key=simon.gerber@vshn2021 -var domain=c-exo-ocp4-test.appuio-beta.ch -var node_state=Running
exoscale_domain.dom: Refreshing state... [id=c-exo-ocp4-test.appuio-beta.ch]
exoscale_compute.vm: Refreshing state... [id=eabb91be-5388-4bbd-9bcc-3e16d062b428]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place

Terraform will perform the following actions:

  # exoscale_compute.vm will be updated in-place
  ~ resource "exoscale_compute" "vm" {
        id                 = "eabb91be-5388-4bbd-9bcc-3e16d062b428"
        name               = "test-vm"
      ~ state              = "Stopped" -> "Running"
        tags               = {}
        # (16 unchanged attributes hidden)
    }

  # exoscale_domain_record.vm[0] will be created
  + resource "exoscale_domain_record" "vm" {
      + domain      = "c-exo-ocp4-test.appuio-beta.ch"
      + hostname    = (known after apply)
      + id          = (known after apply)
      + name        = "test-vm"
      + prio        = (known after apply)
      + record_type = "A"
      + ttl         = 60
    }

Plan: 1 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

exoscale_compute.vm: Modifying... [id=eabb91be-5388-4bbd-9bcc-3e16d062b428]
exoscale_compute.vm: Still modifying... [id=eabb91be-5388-4bbd-9bcc-3e16d062b428, 10s elapsed]
exoscale_compute.vm: Modifications complete after 19s [id=eabb91be-5388-4bbd-9bcc-3e16d062b428]

Error: Provider produced inconsistent final plan

When expanding the plan for exoscale_domain_record.vm[0] to include new values
learned so far during apply, provider
"registry.terraform.io/exoscale/exoscale" produced an invalid new value for
.content: was cty.StringVal(""), but now cty.StringVal("159.100.253.137").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

```

Note that the field `content` for the `exoscale_domain_record` resource is missing in the preview of the changes and that the provider doesn't signal that the `exoscale_compute` resource's field `ip_address` will change.

As far as I can tell, the root cause is that the provider doesn't correctly signal that a `exoscale_compute` resource's attribute `ip_address` may change when it's updated in place.

**Comment by simu** [comment]
Note that I'm provisioning VMs in "Stopped" state initially and powering them on in a subsequent call to `terraform apply` to work around having the VMs come up without a secondary NIC attached (similar to the issue discussed in #48)

**Comment by falzm** [comment]
Hi @simu, I think you're facing this issue because when you run `terraform apply -var node_state=Running`, the Compute instance isn't assigned a public IP address yet, but the `exoscale_domain_record.vm` resource reads the current value for attribute `exoscale_compute.vm.ip_address` from the state (which is first `""`), then during the creation of the DNS record resource the Compute instance gets its public IP address assigned which refreshes the value for the attribute in the state, resulting in the error you're getting.

Can you please try using the actual state of the `exoscale_compute.vm` Compute instance as a predicate for the creation of the DNS record?

```hcl
resource "exoscale_domain_record" "vm" {
  count       = exoscale_compute.vm.state == "Running" ? 1 : 0
  domain      = exoscale_domain.dom.id
  name        = "test-vm"
  ttl         = 60
  record_type = "A"
  content     = exoscale_compute.vm.ip_address
}
```

**Comment by simu** [comment]
Hi @falzm, thanks for the hint, but unfortunately the suggested change results in the same error:

```
$ terraform apply -var ssh_key=simon.gerber@vshn2021 -var domain=c-exo-ocp4-test.appuio-beta.ch -var node_state=Running
exoscale_domain.dom: Refreshing state... [id=c-exo-ocp4-test.appuio-beta.ch]
exoscale_compute.vm: Refreshing state... [id=fa731b60-9cc4-4349-8c84-d17a245c0cbb]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place

Terraform will perform the following actions:

  # exoscale_compute.vm will be updated in-place
  ~ resource "exoscale_compute" "vm" {
        id                 = "fa731b60-9cc4-4349-8c84-d17a245c0cbb"
        name               = "test-vm"
      ~ state              = "Stopped" -> "Running"
        tags               = {}
        # (16 unchanged attributes hidden)
    }

  # exoscale_domain_record.vm[0] will be created
  + resource "exoscale_domain_record" "vm" {
      + domain      = "c-exo-ocp4-test.appuio-beta.ch"
      + hostname    = (known after apply)
      + id          = (known after apply)
      + name        = "test-vm"
      + prio        = (known after apply)
      + record_type = "A"
      + ttl         = 60
    }

Plan: 1 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

exoscale_compute.vm: Modifying... [id=fa731b60-9cc4-4349-8c84-d17a245c0cbb]
exoscale_compute.vm: Still modifying... [id=fa731b60-9cc4-4349-8c84-d17a245c0cbb, 10s elapsed]
exoscale_compute.vm: Modifications complete after 18s [id=fa731b60-9cc4-4349-8c84-d17a245c0cbb]

Error: Provider produced inconsistent final plan

When expanding the plan for exoscale_domain_record.vm[0] to include new values
learned so far during apply, provider
"registry.terraform.io/exoscale/exoscale" produced an invalid new value for
.content: was cty.StringVal(""), but now cty.StringVal("159.100.251.157").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

```

**Comment by falzm** [comment]
If I understand the problem correctly, I'm afraid we won't be able to solve this issue in the provider: it seems the Terraform engine assigns the `cty.StringVal("")` value to the `exoscale_domain_record.vm.content` resource attribute right during the parsing of the configuration graph – as we can see that the prepared plan already reports `~ state = "Stopped" -> "Running"` –, so even if the Compute instance is not running *yet* the DNS record `content` value is already set from the start, then during a subsequent state refresh the provider returns the updated Compute instance IP address value from the API which triggers this error.

I'm sorry but I don't see how to overcome this technical limitation from the perspective of the Exoscale TF provider. Running `terraform apply` once again after the error works, as the state of the Compute instance is stabilized: that's the only workaround I can offer at this point.

**Comment by simu** [comment]
I spent some time reading up on the Terraform provider SDK and found that there's the option to customize the diff logic for resources, cf. https://www.terraform.io/docs/extend/resources/customizing-differences.html

I've added a function which ensures the VM's `ip_address` gets recomputed if the old value of the `ip_address` attribute is `""` and the VM's state is changing from "Stopped" to "Running" which resolves the issue when testing locally. I've submitted #102 which contains this change.
